cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -228,6 +229,48 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
 	return command
 }
 
+// calcNodeCidrSize determines the size of the subnets used on each node, based
+// on the pod subnet provided.  For IPv4, we assume that the pod subnet will
+// be /16 and use /24. For IPv6, there are some restrictions. First, the pod
+// subnet size must be /66 or larger. Second, the bit mask used for allocating
+// subnets (one per node) allows up to 32K (16 bits). To accomodate this, the
+// following algorithm is used, based on pod subnet size (implying the number of
+// bits available), to determine the node CIDR size.
```

Comment:
Doesn't this need to all feed into the pod-per-node configuration?

Reply:
Can you elaborate? Is there some additional work needed? If so, what (as I'm not that familiar with this code)?

From what I can tell, in the existing code, for IPv4, it hard codes the node CIDR to /24 giving 256 pods per node.

This logic that I've introduced, leaves IPv4 unchanged (/24), and attempts to do the same for IPv6, but accommodating for additional constraints.

It'll assign 256 pods per node, when small subnets are used, but as the subnets get larger, it'll split the allocation of pods per node and nodes for the cluster to give > 256 pods per node. When the subnets are very large, we must limit the number of nodes supported (32K), and give the rest for pods per node.

It's expected that very large subnets will be used in practice, with well more than 256 pods per node allowed (and we must, to not violate the constraint for max number of nodes).

As it stands right now, it is totally broken. No subnet size will work.

Reply:
This change is fine, but we need to jig max pods/node with CIDR, which can be done separately.  I also think upping the default to something sane ~256 seems reasonable and all the defaults would just work.   @yujuhong

Reply:
I didn't get the need of changing the (default) max pods per node. The cap should depend on the machine capacity, as well as the workloads.

Reply:
@yujuhong For IPv4, it is using 256 pods per node. 

For IPv6, I tried to do the same thing, however, when the subnets get larger, we are restricted by the number of nodes that can be supported (32K). So, if the subnet is between /66 and /103, we must allow a larger number of pods per node (128-103=25-16=9, 128-66=62-16=46).

Granted, this would be how many pods are **allowed** per node, but that doesn't mean we need to use them.

Reply:
Changing maximum pods per node is a capacity has other implications, e.g., all resources on the node are sufficient to host 100 pods. I don't think in this case we need to change the default in this case. Users should/can configure their kubelets if they want a higher limit to match the cidr.